### DISCUSSION TITLE
Dynamic default values cannot be shown in `--help`

### DISCUSSION THREAD

**Discussion by zanieb** [fastapi/typer#1317] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import typer

def generate():
    return "User!"

def main(name: str = typer.Argument(default=generate)):
    typer.echo(f"Hello {name}")


if __name__ == "__main__":
    typer.run(main)
```

### Description

Running this produces

```
❯ python example.py
Hello User!
```

And help produces
```
❯ python example.py --help
Usage: example.py [OPTIONS] [NAME]

Arguments:
  [NAME]  [default: (dynamic)]
```

However, I'd like to be able to show the default value still. In my use-case, the default is not a constant, but it is deterministic based on user's environment.

I can use `show_default` to set a string e.g. `typer.Argument(default=generate, show_default="test")` but since this cannot be a callable I cannot make it match the value returned by `generate`.

### Wanted Solution

You should be able to override help for dynamic defaults to be similarly dynamic.

### Wanted Code

```python
import typer

def generate():
    return "User!"

def generate_for_help():
    return f"{generate()} (generated)"

def main(name: str = typer.Argument(default=generate, show_default=generate_for_help)):
    typer.echo(f"Hello {name}")


if __name__ == "__main__":
    typer.run(main)
```

### Alternatives

I've explored using `show_default` to provide a string value directly, but it must be a callable to be resolved at command call time rather than at module import time.

Another alternative is to display the dynamic default value instead of `"(dynamic)"` by default. Encourage the user to use `show_default="(dynamic)"` if it's random. This is a breaking change and does not allow for appending additional information to the generated default.

### Operating System

macOS

### Typer Version

0.4.0

### Python Version

3.8

**Comment by zanieb** [comment]
I've implemented this feature in #355, if anyone can think of a workaround please let me know. It is very challenging to customize the output of help since the click classes are obscured.

**Comment by YuriiMotov** [comment]
You can achieve this with f-string (`show_default=f"{generate()} (generated)"`):
```py
import typer

def generate():
    return "User!"

def main(
    name: str = typer.Argument(
        default=generate, show_default=f"{generate()} (generated)"
    ),
):
    typer.echo(f"Hello {name}")

if __name__ == "__main__":
    typer.run(main)

```

**Reply by zanieb** [reply]
Sorry, but that's not what's being requested here.

Specifically

> but it must be a callable to be resolved at command call time rather than at module import time.

The example you provided templates the f-string at module import time. It's equivalent to using a hard-coded string...

e.g.,

```python
BAR = "bar"


def bar() -> str:
    return BAR


class Example:
    def __init__(self, value) -> None:
        self.value = value


def foo(example=Example(f"test {bar()}")):
    print(example.value)


if __name__ == "__main__":
    foo()
    BAR = "not-bar"
    foo()
```

```
❯ uvx python example.py
test bar
test bar
```

**Reply by YuriiMotov** [reply]
Oh, I see.. I missed that requirement.

But, what is your use case?

> In my use-case, the default is not a constant, but it is deterministic based on user's environment.

I understood this as if you need to show different default in help depending on some config parameter (from config file or from env). In this case you can add this condition inside `generate()`.

I wonder why you may need to modify variable inside `if __name__ == "__main__":` before calling the app?

**Reply by zanieb** [reply]
It was a while ago (over 3 years now), and I unfortunately do not remember the details of my specific use-case:

I think the crux is 

> However, I'd like to be able to show the default value still. In my use-case, the default is not a constant, but it is deterministic based on user's environment.
>
> I can use show_default to set a string e.g. typer.Argument(default=generate, show_default="test") but since this cannot be a callable I cannot make it match the value returned by generate

If I have a dynamic `default` value it's generated at execution time, I need my `show_default` value to be calculated at the same time.

> I wonder why you may need to modify variable inside if __name__ == "__main__": before calling the app?

This wasn't intended to be realistic. However, for example, an application could do things like read defaults from a settings file on disk, which you wouldn't want to do at module-load time but you would want to do at startup of the application.

**Reply by YuriiMotov** [reply]
Now I see this makes sense..

I think the feature request should sound something like: **Allow passing callable in `show_default` parameter to make it configurable at run time (e.g. reading possible values from config file)**
